File: 33.투 포인터/2470.두 용액/main.py
import sys

input = sys.stdin.readline

if __name__ == "__main__":
    N = int(input())
    list = sorted(list(map(int, input().split())))

    left, right = 0, N - 1
    result = (float("inf"), 0, 0)
    while left < right:
        cur = list[left] + list[right]

        # 특성값 업데이트
        if abs(cur) < abs(result[0]):
            result = (cur, list[left], list[right])

        # left or right 업데이트
        if cur > 0:
            right -= 1
        else:
            left += 1

    print(result[1], result[2])

File: 33.투 포인터/2470.두 용액/Main.java
import java.io.*;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);

        int left = 0, right = N - 1;
        int[] result = new int[] { Integer.MAX_VALUE }; // 특성값, 용액1, 용액2
        while (left < right) {
            int cur = arr[left] + arr[right];
            if (Math.abs(cur) < Math.abs(result[0])) {
                result = new int[] { cur, arr[left], arr[right] };
            }

            if (cur > 0) {
                right--;
            } else {
                left++;
            }
        }

        bw.write(result[1] + " " + result[2]);
        bw.close();
        br.close();
    }
}